### ISSUE TITLE
Adding firewall rules in Build 15063

### ISSUE THREAD

**Issue by pnegry** [Microsoft/BashOnWindows#1852]
Starting a new issue cause this is a new build...

How does one add rules to windows firewall to permit Pico processes now?

Without a rule to permit processes I get:

Err http://archive.ubuntu.com trusty Release.gpg
  Could not resolve 'archive.ubuntu.com'

**Comment by fpqc** [comment]
delete /etc/resolv.conf and retry

**Comment by stehufntdev** [comment]
Thanks for reporting the issue.  I would agree with @fpqc and try that first.

For the other question above, we've reached out to the Windows firewall team and 3rd party firewalls about adding support for WSL processes.  No ETA on that support but we are actively tracking the issue.

**Comment by pnegry** [comment]
@fpqc I tried that and now it just hangs, so I added the entry for archive.ubuntu.com to /etc/hosts and now I get Cannot initiate the connection to archive.ubuntu.com:80 (91.189.88.161). - connect (13: Permission denied)

The actual issue is that there is still no way to add a rule for a pico process to windows firewall, and as @stehufntdev implies it seems they failed to make any progress on this since RS1.

**Comment by stehufntdev** [comment]
@pnegry the driver API support for WSL processes were published in the creator's update so it will take  the Windows firewall team and 3rd party firewalls time to update.

**Comment by pnegry** [comment]
@stehufntdev so the earliest we might expect support from windows firewall is RS3?

**Comment by stehufntdev** [comment]
I wouldn't expect this change to make the servicing bar for the Creator's update, but that is ultimately the call of the Windows firewall team.  It would help if you can give us feedback on the user voice page so we can track the number of requests to give other teams visibility into the impact - https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/..

**Comment by rodalpho** [comment]
Whelp, didn't make RS3. Any ETA on this being fixed? I'm not going to disable my firewall when I can just spin up a VM.

**Comment by ViRb3** [comment]
Any updates? It's been quite a while now...

**Comment by phpcitizen** [comment]
Still no update?

**Comment by optimisticninja** [comment]
I have done everything in my power to try and get pico processes through my no inbound/only allowed outbound firewall and this is quite frustrating. I first mentioned this issue to an employee on reddit over a year ago and he said the fix was coming in the next release. Even adding source: ephemeral ports -> localhost:53 for all processes (the only non process specific rule in my firewall - you forced my hand) does not work which is logged in my firewall log. An update of any kind would be appreciated.

**Comment by sunilmut** [comment]
The WSL team has been talking to the firewall team about this requirement, but, unfortunately there is no other update to provide at this time. We understand that this is a frustrating limitation, but, some of these dependencies are outside the immediate WSL team and we are working with the right set of teams to help them understand the scenario, requirement etc..Meanwhile, if you haven't upvoted the user voice [ticket](https://wpdev.uservoice.com/forums/266908-command-prompt-console-windows-subsystem-for-l/suggestions/18918169-windows-firewall-should-support-pico-wsl) yet, then please do so. It helps us build the case.

@optimisticninja - Punching a blank inbound rule should work (I use it all the time). Can you describe what is that you are doing? What is your firewall setting? What is not working (what firewall logs are you referring to)? Which Windows build are you on (output of the `ver` command)?

**Comment by optimisticninja** [comment]
These are logs from a frontend I use called Windows Firewall Control 5 (it has a feature that disables windows from creating firewall rules for store apps, etc. automatically). This is a day old installation, so it isn't configured yet via group policy. However - you can see the blocks here.

![capture](https://user-images.githubusercontent.com/21355332/35130836-29c39472-fc80-11e7-8338-70f47a03b014.PNG)

I was trying to only open the ephemeral range to port 53, but I didn't have luck doing this. I can only get WSL picos to work if I allow ALL outbound connections. Do you have a slightly stricter workaround for me in the meantime?

**Comment by sunilmut** [comment]
UDP port 53? Are you trying to get DNS working for WSL processes? Can you also please answer the other questions I have posted before?

Additionally, can you share additional details of the "open the ephemeral range to port 53"? What does it look like?

**Comment by optimisticninja** [comment]
I am not trying to get DNS working for WSL processes. You need to query nameservers...

Output of ver: 

```Microsoft Windows [Version 10.0.16299.192]```

[What is an ephemeral port?](https://en.wikipedia.org/wiki/Ephemeral_port)

Firewall Rules:

Local ports: 49152 to 65535
Destination Address: 192.168.1.1
Destination Port: 53

Which is required prior to the 80 and 443 rules. Just ignore the ephemeral range, you need all three ports punched for a working configuration for basic functionality without getting into other services. I am trying to find a stricter way of allowing WSL without wide open ranges for all processes. I work in a hardened environment. Can I restrict these same settings to the LXSS manager service?

**Comment by sunilmut** [comment]
@optimisticninja - Thanks for the details. You have a DNS server running on `192.168.1.1:53` and these outbound queries are DNS queries. As you have already found out, `LXSS manager service` is just a management service and does not has anything to do with WSL process (pico processes). For most default firewall settings, outbound connections like this are not blocked. Your case seems to be advanced, although equally important. Unfortunately, currently, there is no way to specify WSL processes in firewall rules. The only way is to punch blanket holes for the specific port, inbound or outbound.

**Comment by optimisticninja** [comment]
> I am not trying to get DNS working for WSL processes. You need to query nameservers...

There was some sarcasm there. I appreciate your time, can't wait to see this fixed. Thanks for letting me stir the pot.

**Comment by sunilmut** [comment]
Absolutely. We appreciate everyone's patience here as well.

**Comment by therealkenc** [comment]
Before I go writing something long and wrong, one thing:

> Even adding source: ephemeral ports -> localhost:53 for all processes (the only non process specific rule in my firewall - you forced my hand) does not work which is logged in my firewall log.

That's not the rule. The rule would be [49152 to 65535] -> 192.168.0.1:53. Or maybe [49152 to 65535] -> 8.8.8.8:53.

**Comment by optimisticninja** [comment]
That was a typo in the original post that was fixed hours before you even wrote this, so you wrote something short and wrong. The address at hand was neither of those, if you bothered to look at the logs you would see why the address was 192.168.1.1.

**Comment by therealkenc** [comment]
Whatever dude. I read your log. It wasn't a flame. It was a type-o on your part that you corrected. The longer post (which I didn't want to post until clarification) would have been to Sunil and would have had a nonzero chance of helping your case. I will keep it to myself. Bonne Chance.

**Comment by sunilmut** [comment]
Just an update that Windows firewall integration for WSL processes is on it's way to the insiders. That will bring WSL process integration with firewall to the same level as that of Windows processes. There won't be an ability to add blanket rule to allow WSL process, but, you can manage rules at per process/port level. Hopefully that eases the pain somewhat.

**Comment by DarthSpock** [comment]
@sunilmut I noticed you said "skip-ahead insiders", will the changes made to Windows Defender be documented in some way to allow other A/V, FW vendors to implement something similar? I think this would prove extremely beneficial as there are still many organizations that will 1) won't go to "skip-ahead insiders" or insiders period and/or 2) use another A/V, FW, etc. product such as Avast, Mcafee, Symantec, etc. 

Another important question: Can the changes to Defender be able to be back-ported through normal Windows Updates after a set amount of time of users testing out the new functionality (and resolving any potential bugs)?

**Comment by fpqc** [comment]
@DarthSpock They aren't backporting these changes. The next release will be in around ten months, going by experienxe

**Comment by sunilmut** [comment]
@DarthSpock - As for third party firewall integration, there is a blog post [here](https://blogs.msdn.microsoft.com/wsl/2016/11/01/wsl-antivirus-and-firewall-compatibility/). If anything is missing for their integration, we are happy to assist. Different third-party FW/AV's are at different level of integration with WSL. We have reached out the vendors previously and rely on the community to push this forward with their providers.

As for backporting, @fpqc mentions it. With such short release cycles, backporting features is not very cost effective, generally.

**Comment by DarthSpock** [comment]
Avast appears to hate even the connections to download certain packages even for Kali (probably because those specific packages are used for pen-testing but it isn't all pen-test specific packages interestingly enough). For pico processes, I've had to basically exempt the entire distro (any distro) in the same manner as needed for current Defender in Avast as well. Though I have noticed for a quite awhile now Avast isn't very focused on Linux as much and offers far less features than on Windows (who they clearly target). This will probably be a major change in how they do business so there's now Linux processes running on Windows. I will try pinging them about this again. I'll let you know if they would like assistance aside from what's the in blog. Your implementation for Defender sounds ideal for all vendors so hopefully they can down that route.

**Comment by fpqc** [comment]
@sunilmut Did Microsoft speed up the build calendar? I don't have skip ahead, but the version code seems to say 1809, so September? 1703, 1709, 1803, 1809?

**Comment by rodalpho** [comment]
Good to see this is finally being fixed, even if it took a couple years. Pity it isn't making next month's release.

**Comment by sunilmut** [comment]
@fpqc - Windows is targeting two-feature releases a year. Check out the details section of [this](https://blogs.windows.com/business/2017/04/20/windows-office-align-feature-release-schedules-benefit-customers/#ujegyf2kb43GMjUu.97) blog post.

**Comment by aaronsvk** [comment]
I am unable to setup firewal rules neither for WSL nor for native command line tools. After running following commands, all outbound connections is still allowed from cmd.exe, including `ping google.sk`.
```
netsh advfirewall firewall add rule name="CMD" program="C:\Windows\System32\cmd.exe" dir=out action=block
netsh advfirewall firewall add rule name="PING" program="C:\Windows\System32\PING.EXE" dir=out action=block
```
I can't understand, why is windows firewall incompatible with native windows tools.

**Comment by fpqc** [comment]
@aaronsvk are you running skip ahead?

**Comment by sunilmut** [comment]
@aaronsvk - Try with all lower case on the program name and see if that works. Something like `C:\windows\system32\cmd.exe` or `c:\windows\system32\cmd.exe`

As for WSL integration with firewall, that is coming soon, see the post [above](https://github.com/Microsoft/WSL/issues/1852#issuecomment-371605730).

As far as firewall integration for native Windows tools, that is outside our direct team. I don't know if there is a direct support channel for the firewall. You can also try stackoverflow or other online support channels.

**Comment by Brian-Perkins** [comment]
Better Windows firewall support is now live in the Skip-ahead [17627](https://github.com/MicrosoftDocs/WSL/releases/tag/17627) build.

**Comment by Biswa96** [comment]
How to allow whole WSL system in firewall if all outbound and inbound connections are blocked?

**Comment by sunilmut** [comment]
@Biswa96 - There is no stock policy for WSL in firewall. The level of support for WSL processes in the firewall is the same as that for Windows processes. See the [previous](https://github.com/Microsoft/WSL/issues/1852#issuecomment-371605730) post for more details.

**Comment by rodalpho** [comment]
To be clear, WSL processes (pico) can be firewalled in 17627 and later, and will not be supported in the 1803 major release. So it'll probably be in 1809, coming out in October of this year.

**Comment by fpqc** [comment]
@sunilmut Just wondering, any chance when we list processes in task manager/firewall gui if the pico processes could have a nonblank icon? Sort of a cosmetic thing, but could be nice.

**Comment by sunilmut** [comment]
@fpqc - Agreed that it would be a nice makeup. I will put it in the backlog. Not sure when we will get to it.

**Comment by jerrdale** [comment]
Hi guys, sorry to join the conversation so late. I had this issue with Kali using WSL and WFC. I edited my hosts file to include 192.99.200.113 http.kali.org. (yes, that's spelled right). This got me past the resolution issue, but still could not connect to kali.download on port 80.
I edited the blocked rules in WFC so that the connection would pass.

[Pico rules Partial policy 28.10.2018.zip](https://github.com/Microsoft/WSL/files/2522908/Pico.rules.Partial.policy.28.10.2018.zip)

Basically, I edited the blocked rules in WFC to be **ALL programs,** not just Pico, and **ALL ports** on the originating protocol, left everything else alone.

Hope this helps..

**Comment by aimlessadam** [comment]
@Rodalpho, did firewalling pico processes make it into 1809? 

Thanks!

**Comment by aimlessadam** [comment]
Great! Thank you!
